### PR TITLE
#81

### PR DESCRIPTION
**Pull request by jbutcher21**

### CODE REVIEW

File: G2ConfigTool.py [hunk 1]
```diff
@@ -16,6 +16,15 @@
...
 from datetime import datetime
 from shutil import copyfile
+import subprocess
+
+try:
+    import prettytable
+    from prettytable import PrettyTable
+    from prettytable import ALL as PRETTY_TABLE_ALL
+    from prettytable import SINGLE_BORDER, DOUBLE_BORDER, MARKDOWN, ORGMODE
+except:
```
Comment (antaenc) [inline G2ConfigTool.py:26]: This should catch the expected exception:

except (ImportError, ModuleNotFoundError):

File: G2ConfigTool.py [hunk 2]
```diff
@@ -39,6 +48,138 @@
 
 class G2CmdShell(cmd.Cmd, object):
 
+    def do_help(self, *args):
+        if not args[0]:
+            print(textwrap.dedent(f'''\
+
+                Senzing compares records within and across data sources.  Records consist of features and features have attributes.
```
Comment (antaenc) [inline G2ConfigTool.py:55]: Double space

File: G2ConfigTool.py [hunk 3]
```diff
@@ -39,6 +48,138 @@
...
+                For instance, the NAME feature has attributes such as NAME_FIRST and NAME_LAST, even NAME_ORG for an organization.
```
Comment (antaenc) [inline G2ConfigTool.py:56]: Change even to and

File: G2ConfigTool.py [hunk 4]
```diff
@@ -39,6 +48,138 @@
...
+
+                Features are standardized, expressed in various ways to create candidate keys, and compared to see how close they
```
Comment (antaenc) [inline G2ConfigTool.py:58]: standardized and expressed

File: G2ConfigTool.py [hunk 5]
```diff
@@ -39,6 +48,138 @@
...
+                actually are.
+
+                Finally, a set of rules or "principles" are applied to each candidate record's scores to see if the incoming record should resolve
```
Comment (antaenc) [inline G2ConfigTool.py:61]: candidate should be plural

File: G2ConfigTool.py [hunk 6]
```diff
@@ -39,6 +48,138 @@
...
+                to an existing entity or become a new one. In either case, the rules are also used to create relationships between entities.
+
+                This utility allows you to configure a Senzing instance. Type:
+                    help basic      <- for commonly used commands
```
Comment (antaenc) [inline G2ConfigTool.py:65]: Remove the < use only -

File: G2ConfigTool.py [hunk 7]
```diff
@@ -39,6 +48,138 @@
...
+                    help features   <- to be used only with the guidance of Senzing support
+                    help principles <- to be used only with the guidance of Senzing support
+                    help all        <- to show all configurion commands
+
+                To understand more about configuring Senzing, please review ...
```
Comment (antaenc) [inline G2ConfigTool.py:70]: No ..., use : instead

File: G2ConfigTool.py [hunk 8]
```diff
@@ -48,13 +189,16 @@ def __init__(self, g2module_params, hist_disable, force_mode, file_to_process):
         cmd.Cmd.__init__(self)
 
         # Cmd Module settings
-        self.intro = ''
+        self.intro = '\nType help or ? to list commands.\n'
```
Comment (antaenc) [inline G2ConfigTool.py:192]: Change this back to ='', this message is displayed in the preloop function which is called upon startup and when restarting due to a config change. Now it duplicates the same message:

```
--> ./G2ConfigTool_butcher.py 

Initializing Senzing engines...

Welcome to G2Config Tool. Type help or ? to list commands.


Type help or ? to list commands.
```

File: G2ConfigTool.py [hunk 9]
```diff
@@ -39,6 +48,138 @@
...
+                    addDataSource
+                    deleteDataSource
+                    listDataSources
+
+            While adding or updating features, expressions, scoring thresholds and rules are discouraged without the guidance of Senzing support,
+            knowing how they are configured and what their thresholds are can help you understand why records resolved or not, leading to the
+            proper course of action when working with Senzing Support.
+
+                Features and attribute settings:
+                    listFeatures table
```
Comment (antaenc) [inline G2ConfigTool.py:94]: table is optional consider calling this out or surrounding it in [ ] to highlight it's optional

File: G2ConfigTool.py [hunk 10]
```diff
@@ -39,6 +48,138 @@
...
+
+            NOTE: When you see a how or a why screen output in Senzing, you see the actual entity counts and scores of a match. The list functions
+                above show you what those thresholds and scores are currently configured to.
+
+            '''))
+
+        elif args[0].upper() in ('FEATURE', 'FEATURES'):
+            print(textwrap.dedent(f'''\
+
+            New features and their attributes are rarely needed.  But when they are they are usually industry specific
```
Comment (antaenc) [inline G2ConfigTool.py:111]: Double spacing

File: G2ConfigTool.py [hunk 11]
```diff
@@ -39,6 +48,138 @@
...
+            identifiers (F1s) like medicare_provider_id or swift_code for a bank.  If you want some other kind of attribute like a grouping (FF)
```
Comment (antaenc) [inline G2ConfigTool.py:112]: Double spacing

File: G2ConfigTool.py [hunk 12]
```diff
@@ -39,6 +48,138 @@
...
+            or a physcal attribute (FME, FMES), it is best to clone an existing feature by doing a getFeature, then modifying the json payload to
+            use it in an addFeature.
+
+                Here are the commands to add or update features ...
+                    listFeatures        <- to list all the features in the system
```
Comment (antaenc) [inline G2ConfigTool.py:117]: Change to -

File: G2ConfigTool.py [hunk 13]
```diff
@@ -39,6 +48,138 @@
...
+                    setFeature          <- to change a setting on an existing feature like ...
+                                            - whether or not it is used for candidates
+                                            - what its behavior is when evalated by the rules
+                    deleteFeature       <- to delete a feature added by mistake
+
+
+            Attributes are what you map your source data to.  If you add a new feature, you will also need to add attributes for it. Be sure to
+            use a unique ID for attributes and to classify them as either an ATTRIBUTE or an IDENTIFIER.
+
+                Here are the commands to add or update attributes ...
```
Comment (antaenc) [inline G2ConfigTool.py:129]: : not ...

File: G2ConfigTool.py [hunk 14]
```diff
@@ -39,6 +48,138 @@
...
+                    listAttributes
+                    getAttribute
+                    addAttribute
+                    deleteAttribute
+
+            Some templates have been created to help you add new identifiers if needed. A template adds a feature and its required
+            attributes with one command.
+
+                Here are the commands for using templates
```
Comment (antaenc) [inline G2ConfigTool.py:138]: Add a : to end to be consistent

File: G2ConfigTool.py [hunk 15]
```diff
@@ -39,6 +48,138 @@
...
+                    templateAdd         <- add an identifier (F1) feature and attributes based on a template
```
Comment (antaenc) [inline G2ConfigTool.py:139]: - not <

File: G2ConfigTool.py [hunk 16]
```diff
@@ -39,6 +48,138 @@
...
+        elif args[0].upper() in ('PRINCIPLE', 'PRINCIPLES'):
+            print(textwrap.dedent(f'''\
+
+            Before the principles are applied, the features and expressions created for incoming records are used to find candidates.
+            An example of an expression is name and DOB and there is an expression call on the feature "name" to automatically create it
+            if both a name and DOB are present on the incoming record.  Features and expressions used for candidates are also referred
+            to as candidate builders or candidate keys.
+
+                These are the commands that help with configuring candidate keys:
+                    listFeatures            <- to see what features are used for candidates
```
Comment (antaenc) [inline G2ConfigTool.py:153]: - not <

File: G2ConfigTool.py [hunk 17]
```diff
@@ -39,6 +48,138 @@
...
+                    setFeature              <- to toggle whether or not a feature is used for candidates
+                    listExpressionCalls     <- to see what expressions are currently being created
+                    addToNamehash           <- to add to the name and ... keys
```
Comment (antaenc) [inline G2ConfigTool.py:156]: Why the ...?

File: G2ConfigTool.py [hunk 18]
```diff
@@ -39,6 +48,138 @@
...
+                    addExpressionCall       <- to add a new expression call, aka candidate key
+                    listGenericThresholds   <- to see when candidate keys will become generic and are no longer used to find candidates
+                    setGenericThreshold     <- to change when features with certain behaviors become generic
+
+                WARNING: The cost of raising generic thresholds is speed. It is always best to keep generic thresholds low and to add new
+                            new expressions instead.  You can add additional candidate keys to the name hash with the addToNameHash call
+                            command above or add new expressions by using the addExpressionCall command above.
```
Comment (antaenc) [inline G2ConfigTool.py:163]: You don't need to say above here for both these commands.

File: G2ConfigTool.py [hunk 19]
```diff
@@ -39,6 +48,138 @@
...
+
+            Once the candidate matches have been found, scoring and rule evaluation takes place next.  Scores are rolled up by behavior.
```
Comment (antaenc) [inline G2ConfigTool.py:165]: Remove the next, it's implied.

File: G2ConfigTool.py [hunk 20]
```diff
@@ -39,6 +48,138 @@
...
+            For instance, both addresses and phones have the behavior FF (Frequency Few). So if they both score above their scoring
```
Comment (antaenc) [inline G2ConfigTool.py:166]: Remove So

File: G2ConfigTool.py [hunk 21]
```diff
@@ -39,6 +48,138 @@
...
+            function's close threshold, there would be two CLOSE_FFs (a fragment) which can be used in a rule such as NAME+CLOSE_FF.
```
Comment (antaenc) [inline G2ConfigTool.py:167]: function not function's

File: G2ConfigTool.py [hunk 22]
```diff
@@ -39,6 +48,138 @@
...
+
+                These are the commands that help with configuring rules and scoring:
+                    listRules               <- these are the rules that
```
Comment (antaenc) [inline G2ConfigTool.py:170]: No < just -

File: G2ConfigTool.py [hunk 23]
```diff
@@ -266,7 +410,7 @@ def help_History(self):
 
                 - histClear = Clears the current working session history and the history file. This deletes all history, be careful!
                 - histDedupe = The history can accumulate duplicate entries over time, use this to remove them
-                - histShow = Display all history
+                - history = Display all history
```
Comment (antaenc) [inline G2ConfigTool.py:413]: If changing this to history, change histClear and histDedupe to historyClear and historyDedupe too. Also change in G2Command to be consistent.

File: G2ConfigTool.py [hunk 24]
```diff
@@ -685,7 +829,7 @@ def do_importFromFile(self, arg):
                 return
 
         try:
-            self.cfgData = json.load(open(arg, encoding="utf-8"))
+            self.cfgData = json.load(open(arg), encoding="utf-8")
```
Comment (antaenc) [inline G2ConfigTool.py:832]: This is incorrect and should be: self.cfgData = json.load(open(arg, encoding="utf-8"))
There was a backwards breaking change in the API in Python3.9 that results in: 

```
(g2cfg) exportToFile /tmp/config

Successfully exported!

(g2cfg) importFromFile /tmp/config
ERROR: __init__() got an unexpected keyword argument 'encoding'
  File "/home/ant/senzprojs/3.1.2.22182/python/G2ConfigTool_butcher.py", line 306, in cmdloop
    cmd.Cmd.cmdloop(self)

  File "/usr/lib/python3.9/cmd.py", line 138, in cmdloop
    stop = self.onecmd(line)

  File "/usr/lib/python3.9/cmd.py", line 217, in onecmd
    return func(arg)

  File "/home/ant/senzprojs/3.1.2.22182/python/G2ConfigTool_butcher.py", line 832, in do_importFromFile
    self.cfgData = json.load(open(arg), encoding="utf-8")

  File "/usr/lib/python3.9/json/__init__.py", line 293, in load
    return loads(fp.read(),

  File "/usr/lib/python3.9/json/__init__.py", line 359, in loads
    return cls(**kw).decode(s)
```

File: G2ConfigTool.py [hunk 25]
```diff
@@ -814,19 +958,39 @@ def do_addConfigSectionField(self, arg):
 # ===== data Source commands =====
 
     def do_listDataSources(self, arg):
-        '\n\tlistDataSources [search_value]\n'
+        '''
+        Syntax:
+            listDataSources <optionalFilter> <optionalParameter>
```
Comment (antaenc) [inline G2ConfigTool.py:963]: Optional should be surrounded with [ ]

File: G2ConfigTool.py [hunk 26]
```diff
@@ -814,19 +958,39 @@ def do_addConfigSectionField(self, arg):
...
 
-        print()
+        Examples:
+            listDataSources             <-to list all dataSources in json format
```
Comment (antaenc) [inline G2ConfigTool.py:966]: - not <

File: G2ConfigTool.py [hunk 27]
```diff
@@ -814,19 +958,39 @@ def do_addConfigSectionField(self, arg):
...
+            listDataSources table       <-to list all dataSources in a table format
```
Comment (antaenc) [inline G2ConfigTool.py:967]: These don't need to be prefixed with to, remove

File: G2ConfigTool.py [hunk 28]
```diff
@@ -1053,60 +1227,100 @@ def do_deleteEntityType(self, arg):
...
             else:
                 printWithNewLines('%s rows deleted!' % deleteCnt, 'B')
 
-# ===== feature commands =====
-
     def do_listFunctions(self, arg):
-        '\n\tlistFunctions [search_value]\n'
+        '''
+        Syntax:
+            listFunctions <optionalFilter> <optionalParameter>
```
Comment (antaenc) [inline G2ConfigTool.py:1237]: Optionals in [ ]

File: G2ConfigTool.py [hunk 29]
```diff
@@ -1053,60 +1227,100 @@ def do_deleteEntityType(self, arg):
...
 
-        print()
+        Examples:
+            listFunctions             <-to list all functions in json format
```
Comment (antaenc) [inline G2ConfigTool.py:1240]: no < just -

File: G2ConfigTool.py [hunk 30]
```diff
@@ -1053,60 +1227,100 @@ def do_deleteEntityType(self, arg):
...
-        '\n\tlistFeatures [search_value]\n'
-        # '\n\tlistFeatures\t\t(displays all features)\n' #\
-        # 'listFeatures -n\t\t(to display new features only)\n'
+        '''
+        Syntax:
+            listFeatures <optionalFilter> <optionalParameter>
 
-        print()
+        Examples:
+            listFeatures             <-to list all features in json format
```
Comment (antaenc) [inline G2ConfigTool.py:1300]: You know this one :)

File: G2ConfigTool.py [hunk 31]
```diff
@@ -1716,9 +1947,19 @@ def do_setFeature(self, arg):
...
-        \n\n\taddFeature {"feature":"testFeat", "behavior":"FM", "comparison":"exact_comp", "elementlist": [{"compared": "Yes", "expressed": "No", "element": "test"}]}
-        \n\n\tFor additional example structures, use getFeature or listFeatures\n
+        Syntax:
+            addFeature <json configuration>
+
+        Notes:
+
+            The best way to add a feature is to do a templateAdd as it adds both the feature and its attributes.
+
+            If you need to add one manually, it is best to clone an existing one by doing a getFeature on one like it,
```
Comment (antaenc) [inline G2ConfigTool.py:1957]: If you need to add a feature manually, it is recommended to clone an existing and similar one with getFeature

File: G2ConfigTool.py [hunk 32]
```diff
@@ -1716,9 +1947,19 @@ def do_setFeature(self, arg):
...
+            and editing its json configuration in your favorite editor.  You will need to at least change its "id" and
```
Comment (antaenc) [inline G2ConfigTool.py:1958]: and editing the JSON. You will need to at least change "id" and

File: G2ConfigTool.py [hunk 33]
```diff
@@ -1716,9 +1947,19 @@ def do_setFeature(self, arg):
...
+            "feature" code so they are unique, but you can change anything else you like.  Then when you are ready, use
```
Comment (antaenc) [inline G2ConfigTool.py:1959]: "feature" codes to be unique, you can change anything else required. When ready, use

File: G2ConfigTool.py [hunk 34]
```diff
@@ -1716,9 +1947,19 @@ def do_setFeature(self, arg):
...
+            it as the addRecord <json configuration>.
```
Comment (antaenc) [inline G2ConfigTool.py:1960]: the modified JSON with addRecord <json configuration>.

File: G2ConfigTool.py [hunk 35]
```diff
@@ -1716,9 +1947,19 @@ def do_setFeature(self, arg):
...
+
+            And don't forget to add the attributes for its elements using the addAttribute command!
```
Comment (antaenc) [inline G2ConfigTool.py:1962]: Remove And

File: G2ConfigTool.py [hunk 36]
```diff
@@ -2052,7 +2293,18 @@ def getFeatureJson(self, ftypeRecord):
...
-        '\n\taddToNamehash {"feature": "<feature>", "element": "<element>"}\n'
+        '''
+        Syntax:
+            addToNamehash {"feature": "<feature>", "element": "<element>"}
+
+        Example:
+            addToNamehash {"feature": "ADDRESS", "element": "STR_NUM"}
+
+        Notes:
+            This command appends an attribute from another feature to the name hash.  In the example above, the street number
```
Comment (antaenc) [inline G2ConfigTool.py:2304]: Double space

File: G2ConfigTool.py [hunk 37]
```diff
@@ -2294,29 +2546,37 @@ def do_deleteFromSSNLast4hash(self, arg):
...
     def do_getAttribute(self, arg):
         '''
-        \n\tgetAttribute {"attribute": "<attribute_name>"}
-        \n\tgetAttribute {"feature": "<feature_name>"}\t\tList all the attributes for a feature\n
-        '''
+        Syntax:
+            getAttribute <attribute>
 
+        Examples:
+            getAttribute NAME_LAST             <- shows attribute json
```
Comment (antaenc) [inline G2ConfigTool.py:2577]: You know this one by now

**Review by antaenc — CHANGES_REQUESTED** [review]
Syntax and small changes mostly to improve readability.

File: G2ConfigTool.py (hunk 1 above)
Comment (jbutcher21) [reply G2ConfigTool.py:26]: Only important if I were gonna report that specifically.   If no pretty table, then no pretty table support.   warning to install it is later if they even try to use it.

File: G2ConfigTool.py (hunk 6 above)
Comment (jbutcher21) [reply G2ConfigTool.py:65]: why?   I much prefer the arrow

File: G2ConfigTool.py (hunk 18 above)
Comment (jbutcher21) [reply G2ConfigTool.py:163]: not sure what you mean.  I like this.

File: G2ConfigTool.py (hunk 21 above)
Comment (jbutcher21) [reply G2ConfigTool.py:167]: Nope, it is possive

File: G2ConfigTool.py (hunk 22 above)
Comment (jbutcher21) [reply G2ConfigTool.py:170]: disagree

File: G2ConfigTool.py (hunk 23 above)
Comment (jbutcher21) [reply G2ConfigTool.py:413]: not important.   everyone uses history.   They other two are rarely used and can be looked up if needed.

File: G2ConfigTool.py (hunk 26 above)
Comment (jbutcher21) [reply G2ConfigTool.py:966]: disagree

File: G2ConfigTool.py (hunk 27 above)
Comment (jbutcher21) [reply G2ConfigTool.py:967]: disagree

File: G2ConfigTool.py (hunk 29 above)
Comment (jbutcher21) [reply G2ConfigTool.py:1240]: disagree

File: G2ConfigTool.py [hunk 38]
```diff
@@ -1053,60 +1227,100 @@ def do_deleteEntityType(self, arg):
...
+            json_lines.append({"id": fclassRecord['FCLASS_ID'], "class": fclassRecord['FCLASS_CODE']})
+        print_json_lines(json_lines, display_format)
 
     def do_listFeatures(self, arg):
-        '\n\tlistFeatures [search_value]\n'
-        # '\n\tlistFeatures\t\t(displays all features)\n' #\
-        # 'listFeatures -n\t\t(to display new features only)\n'
+        '''
+        Syntax:
+            listFeatures <optionalFilter> <optionalParameter>
```
[the remark this answers is not in the document]
Comment (jbutcher21) [reply G2ConfigTool.py:1297]: only obvious to a programmer.  I prefer as is.

File: G2ConfigTool.py (hunk 28 above)
Comment (jbutcher21) [reply G2ConfigTool.py:1237]: this for end users not programmers.  above is very clear.

**Review by antaenc — APPROVED** [review]

**Pull request merged**